### PR TITLE
Upgrade to SM115

### LINKED ISSUE #345: The Integrity check is currently broken
Integrity checks are failing again due to missing SM pkg, so I've asked on SpiderMonkey Matrix about more permanent SM pkgs and [they said that SM packages are not recommended way of using SM](https://matrix.to/#/!CZEbuMfGYVdQMIBKeP:mozilla.org/$lLzrgflV1kUvMonPfMPgwHU9y34pvFrQ3e4B1ohN-TI?via=mozilla.org&via=matrix.org&via=igalia.com), so they do not plan to host them on ftp anytime soon.

Here are few possible solutions for going forward:
1. Do nothing and keep Integrity check as optional (so commits are merged even if check fails)
2. Use whole mozilla-source for SM
3. Host used SM pkgs somewhere (maybe as GitHub releases)
4. Follow ESR cycle (thus never falling behind for more than 3 months = time that SM pkgs are available)

The last one may be the best for servo as porting ESR patches should not be as hard as usual SMups and following ESR would keep SM properly patched. ESR version are released once per year and they usually come with [Migration Guides](https://github.com/mozilla-spidermonkey/spidermonkey-embedding-examples/blob/esr102/docs/Migration%20Guide.md) and there are other users of SM which are also doing upgrades (most notably [gjs](https://gitlab.gnome.org/GNOME/gjs)) which could help our upgrades. The only problem are slower features rollout, but current servo development may benefit from less SM breaking changes so there could be more focus on other parts.

### PR DESCRIPTION
**Pull request by sagudev**
This PR updates SpiderMonkey to 115.3.0 (ESR).

The main blocker for SM bump is/was [removal of streams support](https://phabricator.services.mozilla.com/D159610), so I wondered how hard would it be to just revert deletion of streams in SM. It was big success, with practically no additional changes needed. This is just short-term solution so we can start tracking esr115 with mozjs (read more in https://github.com/servo/mozjs/issues/345), but streams really do need to be replaced with next ESR (something I will probably do after some more binding fixing in servo).

Fixes #345

Servo companion PR: https://github.com/servo/servo/pull/30379

~~It would be nice if we land both #407 and #402 before this.~~ DONE

NOTE TO REVIEWERS: This PR should be reviewed per commit.
Also it would be good if we could merge this without squashing.

### CODE REVIEW

**Comment by mukilan** [comment]
This is awesome! Regarding Android failure, this is something @mrobinson and I faced as well when working on resurrecting servo's android support.

>   ERROR: Only clang/llvm 13.0 or newer is supported for armv7a-unknown-linux-androideabi30 (found version 12.0.9).

This error happens because Gecko's build system requires Clang 13 but the NDK r21 ships with Clang 12. To make this work we should ideally target NDK r25c, which will also allow us to use the current rustc version since NDK 23 and above don't have libgcc and older rust requires them. There are few other issues that might also need to be fixed that I forget now.

I'll be happy to look into fixing these errors and making android work on this branch tomorrow.

**Comment by sagudev** [comment]
Actually I already put android on r23 as there was error of to old ndk. Newer versions might not work, but I did not really investigate further.

> I'll be happy to look into fixing these errors and making android work on this branch tomorrow.

Great, android isn't really my cup of tea. Branch is currently quite messy so maybe it would make sense to do some cleaning from my end before hand. I will let you know when it is in more solid state.

**Comment by jdm** [comment]
Strong agreement that this path makes sense!

**Comment by sagudev** [comment]
@mukilan it's cleaned, so feel free to work on android.

**Comment by mukilan** [comment]
Thanks @sagudev! I'm looking into the android build issues.

**Comment by mukilan** [comment]
@sagudev I've posted my changes [here]( https://github.com/mukilan/mozjs/tree/sm115-with-streams). The build is passing on CI and locally, but there are a lots of compiler warnings now. I suspect the flags we specify using [cc::Build::flag_if_supported](https://github.com/servo/mozjs/blob/master/mozjs/build.rs#L272) are not being passed along to the compiler. I'm not sure why this is happening, but I'll investigate this separately. 

I've also tested this version of mozjs with servo on android, using my local patches for android  + your PR in servo for updateing mozjs and I didn't see any issues related to mozjs and servo.org works, so I believe SM 115 + NDK r25c should not be a problem.

**Comment by sagudev** [comment]
Thank you!

> The build is passing on CI and locally, but there are a lots of compiler warnings now. I suspect the flags we specify using [cc::Build::flag_if_supported](https://github.com/servo/mozjs/blob/master/mozjs/build.rs#L272) are not being passed along to the compiler. I'm not sure why this is happening, but I'll investigate this separately.

There wasn't any warnings before? I think we should left this for followup, but I think mozilla has set of it's own (no) warnings list so we should just copy those into cc flags.

**Comment by sagudev** [comment]
I think this should be ready now for review.

**Review by jdm — APPROVED** [review]
Woo! I feared there would be many more changes required, so this is a nice surprise.

**Comment by sagudev** [comment]
@mrobinson could you do merge without squash.

**Comment by mrobinson** [comment]
What a great change! I'll take a look at this today.

**Comment by sagudev** [comment]
Just rebase.

File: mozjs/mozjs/js/src/vm/StructuredClone.cpp [hunk 1]
```diff
@@ -1894,13 +1894,13 @@ bool JSStructuredCloneWriter::traverseError(HandleObject obj) {
...
   // The Error stack property is saved as SavedFrames.
   RootedValue stack(cx, NullValue());
-  RootedObject stackObj(cx, unwrapped->stack());
+  /*RootedObject stackObj(cx, unwrapped->stack());
   if (stackObj && stackObj->canUnwrapAs<SavedFrame>()) {
     stack.setObject(*stackObj);
     if (!cx->compartment()->wrap(cx, &stack)) {
       return false;
     }
-  }
```
Comment (mrobinson) [inline mozjs/mozjs/js/src/vm/StructuredClone.cpp]: If this is non-standard, why is it working for Gecko I wonder?

**Comment by sagudev** [comment]
Hm, clang gets mingw target insted of msvc one:
> Executing: `C:/hostedtoolcache/windows/sccache/0.5.4/x64/sccache.exe C:/PROGRA~1/LLVM/bin/clang-cl.exe -fms-compatibility-version=19.37 -Xclang -std=c++17 --target=aarch64-mingw32 'C:\Users\RUNNER~1\AppData\Local\Temp\conftest.mnnm7276.cpp' -c`

**Comment by sagudev** [comment]
https://searchfox.org/mozilla-central/source/build/moz.configure/init.configure#654
```python
    # Config.sub doesn't like the *-windows-msvc/*-windows-gnu triplets, so
    # munge those before and after calling config.sub.
    suffix = None
    munging = {
        "-windows-msvc": "-mingw32",
        "-windows-gnu": "-mingw32",
    }
```

File: mozjs/mozjs/js/src/vm/StructuredClone.cpp (hunk 1 above)
Comment (sagudev) [reply mozjs/mozjs/js/src/vm/StructuredClone.cpp]: Done!

**Comment by sagudev** [comment]
Historically we did pass -mingw32 targets, but these targets are not used anymore. See https://searchfox.org/mozilla-esr115/source/browser/config/mozconfigs/win64-aarch64/common-win64#3 for aarch64 example.

**Comment by mrobinson** [comment]
I plan to land this today, @sagudev, but do you mind if I organize these commits a bit (squashing and editing messages) before landing?

**Comment by sagudev** [comment]
> I plan to land this today, @sagudev, but do you mind if I organize these commits a bit (squashing and editing messages) before landing?

Do as you wish, just keep [Update SpiderMonkey](https://github.com/servo/mozjs/pull/408/commits/f38cb15d0bf8140330a82249d630e8c368e8e9d2) as it is, as it was generated by update.py script.

**Comment by mrobinson** [comment]
> Do as you wish, just keep [Update SpiderMonkey](https://github.com/servo/mozjs/pull/408/commits/f38cb15d0bf8140330a82249d630e8c368e8e9d2) as it is, as it was generated by update.py script.

Okay. I haven't changed the contents, but I've given it a slightly more descriptive commit message.

**Review by mrobinson — APPROVED** [review]

**Comment by mrobinson** [comment]
CI failure looks like simple network flakiness:

```
warning: spurious network error (3 tries remaining): [7] Couldn't connect to server (Failed to connect to index.crates.io port 443 after 4003 ms: Couldn't connect to server)
warning: spurious network error (2 tries remaining): [7] Couldn't connect to server (Failed to connect to index.crates.io port 443 after 0 ms: Couldn't connect to server)
warning: spurious network error (1 tries remaining): [7] Couldn't connect to server (Failed to connect to index.crates.io port 443 after 0 ms: Couldn't connect to server)
error: failed to get `encoding_c` as a dependency of package `mozjs_sys v0.68.2 (/__w/mozjs/mozjs/mozjs)`
```

**Pull request merged**